app/models/concerns/import_job.rb

```diff
@@ -35,7 +35,19 @@ def validate_headers(filename)
 
     log("Headers in file: #{headers}", :debug)
     log("Valid headers for model: #{valid_headers}", :debug)
-    valid_headers == headers
+
+    success = valid_headers == headers
+    unless success
+      ActionCable.server.broadcast "import_job", { html:
+        "<div class='notification is-warning'>
+          <button class='delete'></button>
+          <strong>Invalid headers</strong>
+          <p>Valid headers: #{valid_headers.join(', ')}</p>
+          <p>Current headers: #{headers.join(', ')}</p>
```

Comment:
If you were to go with the JSON route (or really even if you didn't for testing reasons) it may be easier to work with this HTML:


```html
<div class='notification is-warning'>
  <button class='delete'></button>
  
  <div class="notification-error-title">  
    <strong>Error:</strong> <span class="notification-error-type">Invalid Headers</span>
  </div>

  <div class="notification-error-content">
  </div>
</div>
```

...which would make the associated jQuery bind to replacing the content of that something like this:

```javascript
// These two will already make testing easier by being able to pull sections by a specific marker.
const $notificationErrorType = $('.notification-error-type');
const $notificationErrorContent = $('notification-error-content');

const wrap = tagName => contents => `<${tagName}>${Array.from(contents).join('')}</${tagName}>`
const ul = wrap('ul');
const li = wrap('li');

const list = items => ul(items.map(li));

App.notifications = App.cable.subscriptions.create("NotificationsChannel", {
  received ({ invalid_headers, valid_headers }) {
    $notificationErrorContent.html(list(data.invalid_headers));
  }
});
```

Give or take, something like that.

Reply:
That'd also allow you to specify the error type in the JSON response making that a bit more of a universal component that could be used for any type of async error later.